### DISCUSSION TITLE
mount filesystem outside of kubernetes cluster

### DISCUSSION THREAD

**Discussion by caffeinism** [rook/rook#11397] [Q&A] [answered]
I would also like to use rook-ceph on machines outside of kubernetes.

The problem I have is that mon does not belong to a certain node.

1. Manually register in fstab
But I think this method will make the machine unbootable when mon is changed, which will make life hard. 🥲

2. Update and mount endpoints of mon as a service in linux.
I'm afraid another program might need the volume before this script runs.

Nonetheless, I think the latter method works better for me. Are there any other good ways?

**Comment by caffeinism** [comment]
After thinking about it, I realized that I could just proceed with the second method... Sorry for posting without thinking deeply.

**Reply by travisn** [reply]
[Host networking](https://github.com/rook/rook/blob/release-1.10/deploy/examples/cluster.yaml#L91-L92) is generally recommended. Then the mon IPs and other ceph daemons would be addressable outside the cluster.

**Reply by caffeinism** [reply]
But when mon becomes unavailable, won't it deploy to another node? I'm trying to supplement by inserting a script to watch kubernetes just in case. Is this the right way? I referred to toolbox.sh of direct-mount.

**Reply by caffeinism** [reply]
I found that this could be replaced by an A-record domain with multiple IPs assigned to it. This looks relatively neat, but you still have to manually change the records. Is there any way to automate this a bit more? I tried with kubernetes' LoadBalancer type service, but it doesn't work properly. Is there a setting in ceph that specifies the IP to be bound separately?

**Reply by travisn** [reply]
The mons will either be listening on the host ip if host networking is enabled, or on the service ip. Either one will change if the mons are failed over. In the external clusters where rook is running, rook does query the provider cluster (every 60s IIRC) in case the mons are updated, so the consumer clusters will be able to continue working automatically in the mons change. With host networking as long as the hosts are addressable by the consumer clusters, I think you're good to go.

**Reply by caffeinism** [reply]
I was thinking about what to do when a mon dies and the new mon changes to a different host that I didn't specify. I exposed coredns and specified mon as a headless service and mounted it through an internal domain, but it doesn't seem to be good.

**Reply by travisn** [reply]
You could consider adding node affinity to the mons so they only start on the nodes you specify, but then it would limit the ability of mon failover to work when a node fails.

**Comment by caffeinism** [comment, marked as answer]
In the end, I decided to change the configuration of coredns to expose the domain of mon to the outside using the k8s_external function and mount it through it.
